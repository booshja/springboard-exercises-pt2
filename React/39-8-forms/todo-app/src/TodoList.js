import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";

const TodoList = () => {
    // this component should render the NewTodoForm component and should render the list of Todo components. Place your state that contains all of the todos in this component.
    const INITIAL_STATE = [];

    const [todos, setTodos] = useState(INITIAL_STATE);

    const addTodo = (newTodo) => {
        setTodos((todos) => [...todos, { id: uuid(), ...newTodo }]);
    };

    const removeTodo = (id) => {
        setTodos(todos.filter((todo) => todo.id !== id));
    };

    return (
        <div>
            <h1>To Do List</h1>
            <NewTodoForm addTodo={addTodo} />
            <div className="TodoList--TodosContain">
                {todos.map(({ task, id }) => (
                    <Todo
                        id={id}
                        task={task}
                        removeTodo={removeTodo}
                        key={id}
                    />
                ))}
            </div>
        </div>
    );
};

export default TodoList;
