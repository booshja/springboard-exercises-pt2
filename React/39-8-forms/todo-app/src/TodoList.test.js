import React from "react";
import { fireEvent, render } from "@testing-library/react";
import TodoList from "./TodoList";

it("renders without crashing", () => {
    render(<TodoList />);
});

it("matches snapshot", () => {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

it("should add a todo", () => {
    const { getByText, queryByText, getByPlaceholderText } = render(
        <TodoList />
    );
    const taskInput = getByPlaceholderText("New Todo");
    const btn = getByText("Add Todo");

    expect(queryByText("X")).not.toBeInTheDocument();

    fireEvent.change(taskInput, { target: { value: "Walk dogs" } });
    fireEvent.click(btn);

    expect(getByText("X")).toBeInTheDocument();
});

it("should clear inputs after submission", () => {
    const { getByPlaceholderText, getByText } = render(<TodoList />);
    const taskInput = getByPlaceholderText("New Todo");
    const btn = getByText("Add Todo");

    fireEvent.change(taskInput, { target: { value: "Defragment Computer" } });

    fireEvent.click(btn);

    expect(taskInput.value).toEqual("");
});
