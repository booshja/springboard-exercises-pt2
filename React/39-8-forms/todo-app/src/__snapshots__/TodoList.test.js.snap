// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      To Do List
    </h1>
    <form
      class="NewTodoForm"
    >
      <label
        class="NewTodoForm--label"
        for="task"
      >
        New Todo:
      </label>
      <input
        class="NewTodoForm--task"
        id="task"
        name="task"
        placeholder="New Todo"
        type="text"
        value=""
      />
      <button
        class="NewTodoForm--btn"
      >
        Add Todo
      </button>
    </form>
    <div
      class="TodoList--TodosContain"
    />
  </div>
</DocumentFragment>
`;
